File: haskell/glue/src/Glue/Lib/Math/Utility/Ceil.hs
module Glue.Lib.Math.Utility.Ceil where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Ceiling function
-- Mirrors Haskell Glue.Lib.Math.Utility.Ceil.ceil exactly
ceil :: IR Eval
ceil = NativeFunc ceilImpl

-- Ceiling function implementation
-- Mirrors Haskell Glue.Lib.Math.Utility.Ceil.ceilImpl exactly
ceilImpl :: IR Eval -> Eval (IR Eval)
ceilImpl arg = case arg of
    Integer n -> pure $ Integer n
    Float n -> pure $ Integer (Prelude.ceiling n)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/utility/ceil.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Ceiling function (rounds up to nearest integer)
/// Mirrors Haskell Glue.Lib.Math.Utility.Ceil.ceil exactly
final Ir ceil = IrNativeFunc(ceilImpl);

/// Ceiling function implementation (rounds up to nearest integer)
/// Mirrors Haskell Glue.Lib.Math.Utility.Ceil.ceilImpl exactly
Eval<Ir> ceilImpl(Ir arg) {
  return eval(arg).flatMap((va) {
    return switch (va) {
      IrInteger(value: final n) => Eval.pure(IrInteger(n)),
      IrFloat(value: final n) => Eval.pure(IrInteger(n.ceil())),
      _ => throwError(wrongArgumentType(['number'])),
    };
  });
}
